File: src/quest_editor/controllers/QuestInfoController.ts
import { Property } from "../../core/observable/property/Property";
import { EditIdAction } from "../actions/EditIdAction";
import { EditNameAction } from "../actions/EditNameAction";
import { EditShortDescriptionAction } from "../actions/EditShortDescriptionAction";
import { EditLongDescriptionAction } from "../actions/EditLongDescriptionAction";
import { QuestModel } from "../model/QuestModel";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Controller } from "../../core/controllers/Controller";

export class QuestInfoController extends Controller {
    readonly current_quest: Property<QuestModel | undefined>;
    readonly enabled: Property<boolean>;
    readonly unavailable: Property<boolean>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        this.current_quest = store.current_quest;
        this.enabled = store.quest_runner.running.map(r => !r);
        this.unavailable = this.current_quest.map(q => q == undefined);
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    set_id = (id: number): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo.push(new EditIdAction(quest, quest.id.val, id)).redo();
        }
    };

    set_name = (name: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo.push(new EditNameAction(quest, quest.name.val, name)).redo();
        }
    };

    set_short_description = (short_description: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo
                .push(
                    new EditShortDescriptionAction(
                        quest,
                        quest.short_description.val,
                        short_description,
                    ),
                )
                .redo();
        }
    };

    set_long_description = (long_description: string): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.store.undo
                .push(
                    new EditLongDescriptionAction(
                        quest,
                        quest.long_description.val,
                        long_description,
                    ),
                )
                .redo();
        }
    };
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/controllers/QuestInfoController.kt
package world.phantasmal.web.questEditor.controllers

import world.phantasmal.observable.cell.*
import world.phantasmal.web.questEditor.actions.EditPropertyAction
import world.phantasmal.web.questEditor.stores.QuestEditorStore
import world.phantasmal.webui.controllers.Controller

class QuestInfoController(private val store: QuestEditorStore) : Controller() {
    val unavailable: Cell<Boolean> = store.currentQuest.isNull()
    val enabled: Cell<Boolean> = store.questEditingEnabled

    val episode: Cell<String> = store.currentQuest.map { it?.episode?.name ?: "" }
    val id: Cell<Int> = store.currentQuest.flatMap { it?.id ?: cell(0) }
    val name: Cell<String> = store.currentQuest.flatMap { it?.name ?: emptyStringCell() }
    val shortDescription: Cell<String> =
        store.currentQuest.flatMap { it?.shortDescription ?: emptyStringCell() }
    val longDescription: Cell<String> =
        store.currentQuest.flatMap { it?.longDescription ?: emptyStringCell() }

    fun focused() {
        store.makeMainUndoCurrent()
    }

    fun setId(id: Int) {
        if (!enabled.value) return

        store.currentQuest.value?.let { quest ->
            store.executeAction(EditPropertyAction("Edit ID", quest::setId, id, quest.id.value))
        }
    }

    fun setName(name: String) {
        if (!enabled.value) return

        store.currentQuest.value?.let { quest ->
            store.executeAction(
                EditPropertyAction("Edit name", quest::setName, name, quest.name.value)
            )
        }
    }

    fun setShortDescription(shortDescription: String) {
        if (!enabled.value) return

        store.currentQuest.value?.let { quest ->
            store.executeAction(
                EditPropertyAction(
                    "Edit short description",
                    quest::setShortDescription,
                    shortDescription,
                    quest.shortDescription.value,
                )
            )
        }
    }

    fun setLongDescription(longDescription: String) {
        if (!enabled.value) return

        store.currentQuest.value?.let { quest ->
            store.executeAction(
                EditPropertyAction(
                    "Edit long description",
                    quest::setLongDescription,
                    longDescription,
                    quest.longDescription.value,
                )
            )
        }
    }
}
